OrbitGl/TracepointTrack.cpp

```diff
@@ -69,14 +69,38 @@ void TracepointTrack::UpdatePrimitives(uint64_t min_tick, uint64_t max_tick,
 
   const Color kWhite(255, 255, 255, 255);
 
+  const Color kTracepoint(255, 255, 255, 190);
+
   const Color kGreenSelection(0, 255, 0, 255);
 
   if (!picking) {
     for (auto it = tracepoints.lower_bound(min_tick); it != tracepoints.upper_bound(max_tick);
          ++it) {
       uint64_t time = it->first;
       Vec2 pos(time_graph_->GetWorldFromTick(time), pos_[1]);
-      batcher->AddVerticalLine(pos, -track_height, z, kWhite);
+      batcher->AddVerticalLine(pos, -track_height / 4, z, kTracepoint);
+      batcher->AddVerticalLine(Vec2(pos[0], pos[1] - track_height), track_height / 4, z,
+                               kTracepoint);
+      pos[1] -= track_height / 2;
+
+      float x_pos = pos[0];
+      float y_pos = pos[1];
+      float radius = track_height / 4;
+      const float steps = 22;
+      const float angle = (kPiFloat * 2.f) / steps;
+      float prev_x = x_pos;
+      float prev_y = y_pos - radius;
+      for (int i = 1; i <= steps; i++) {
+        float new_x = pos[0] + radius * sinf(angle * i);
+        float new_y = pos[1] - radius * cosf(angle * i);
```

Comment:
It would be better to compute the points of the circle once for all the tracepoints of the track. Even better, we could compute the points of a unit circle only once for the lifetime of the app, then scale those points by "radius" once per track "UpdatePrimitives", and then finally translate those points once per tracepoint. This would avoid performing costly trigonometric functions for every tracepoint on every track.

Reply:
+1

Reply:
Thank you so much! I will update the code!